### PR TITLE
Supporting complex types in ENC (WAS: ENC for tuples in foreach)

### LINKED ISSUE #19303: Support matching complex types between ENC generations
The `ArrayOfTuples` test (in `EditAndContinueTests.cs`) and a number of other tests seem to be creating ValueTuple types (regular NamedTypeSymbol rather than TupleTypeSymbol).
This is unexpected, as the tuple type unification work is supposed to lift any instances of `ValueTuple<...>` to a tuple `(...)`.

Here's a place to set breakpoint: http://source.roslyn.io/#Microsoft.CodeAnalysis/Emit/EditAndContinue/EncLocalInfo.cs,58

### PR DESCRIPTION
**Pull request by ivanbasov**
**Customer scenario**

On ENC, user changes code near a variable declaration of a complex type, e.g. tuple or array of tuples like 
```
var a = (1, 2);
<breakpoint is here> Console.WriteLine(0);
```
to
```
var a = (1, 2);
Console.WriteLine(1);
```

Expected: the variable has a value defined just above.
Actual: the variable value is reseted. Internally, a new variable is declared within ENC without a match to the existing one.

Bugs this fixes:

Fixes #19303

Workarounds, if any

Risk

Low

Performance impact

Low

Is this a regression from a previous update?

Root cause analysis:

How did we miss it? What tests are we adding to guard against it in the future?

How was the bug found?

Planned

### CODE REVIEW

File: src/Compilers/CSharp/Test/Emit/Emit/EditAndContinue/EditAndContinueTests.cs [hunk 1]
```diff
@@ -8077,5 +8077,403 @@ .maxstack  2
 }
 ");
         }
+
+        [Fact]
+        public void ForeachStatement_Reorder()
```
Comment (tmat) [inline src/Compilers/CSharp/Test/Emit/Emit/EditAndContinue/EditAndContinueTests.cs]: >ForeachStatement_Reorder [](start = 20, length = 24)

You can remove this test. Instead expand ArrayOfTuples to test 2nd generation.

Comment (tmat) [inline src/Compilers/CSharp/Test/Emit/Emit/EditAndContinue/EditAndContinueTests.cs]: Also, add ArrayOfTuple test to VB


---
In reply to: [115573899](https://github.com/dotnet/roslyn/pull/19283#discussion_r115573899) [](ancestors = 115573899)

Comment (tmat) [inline src/Compilers/CSharp/Test/Emit/Emit/EditAndContinue/EditAndContinueTests.cs]: Actually... let's make the ArrayOfTuples test more interesting and make the type of the variable something like an array of tuples of anonymous types of a List<tuple>. You can also add more variables of more complex types, including pointers, ref variables, etc.

And rename the test to ComplexTypes, or something like that.

 

---
In reply to: [115574092](https://github.com/dotnet/roslyn/pull/19283#discussion_r115574092) [](ancestors = 115574092,115573899)

Comment (ivanbasov) [inline src/Compilers/CSharp/Test/Emit/Emit/EditAndContinue/EditAndContinueTests.cs]: Thank you! Fixed

---
In reply to: [115576722](https://github.com/dotnet/roslyn/pull/19283#discussion_r115576722) [](ancestors = 115576722,115574092,115573899)

File: src/Compilers/CSharp/Test/Emit/Emit/EditAndContinue/EditAndContinueTests.cs [hunk 2]
```diff
@@ -8202,5 +8202,227 @@ .maxstack  2
...
+using System.Collections.Generic;
+
+class C
+{
+    public static void G()
+    {       
+        var <N:0>a</N:0> = new { key = ""a"", value = new List<(int, int)>()};
+        var <N:1>b</N:1> = (number: 5, value: a);
+        var <N:2>c</N:2> = new[] { b };
+        IntPtr[] <N:3>d</N:3> = new IntPtr[1];
```
Comment (tmat) [inline src/Compilers/CSharp/Test/Emit/Emit/EditAndContinue/EditAndContinueTests.cs]: >IntPtr [](start = 8, length = 6)

This is not really a pointer type. IntPtr is just ordinary value type. 

You can use the following trick to cover quite a few interesting cases:

```C#
class C<T>
. {
.     public enum E
.     {
.         A
.     }
. }


C<(int, dynamic)>.E***[,,] x;
```

**Comment by tmat** [comment]
:shipit:

File: src/Compilers/VisualBasic/Test/Emit/Emit/EditAndContinue/EditAndContinueTests.vb [hunk 3]
```diff
@@ -5285,7 +5287,162 @@ End Class")
...
+        <Fact>
+        Public Sub ComplexTypes()
+            Dim originalSourceCode = "
+Imports System
+Imports System.Collections.Generic
 
+Class C
+    Sub G()
+        Dim <N:0>a</N:0> = New With {.Key = ""a"", .Value = New List(Of Tuple(Of Integer, Integer))()}
+        Dim <N:1>b</N:1> = Tuple.Create(5, a)
```
Comment (tmat) [inline src/Compilers/VisualBasic/Test/Emit/Emit/EditAndContinue/EditAndContinueTests.vb]: >Tuple.Create(5, a) [](start = 27, length = 18)

This doesn't create a value tuple. I think VB has syntax for tuples as well.

File: src/Compilers/VisualBasic/Test/Emit/Emit/EditAndContinue/EditAndContinueTests.vb [hunk 4]
```diff
@@ -5285,7 +5287,162 @@ End Class")
...
+        Dim <N:2>c</N:2> = {b}
+        Dim <N:3>d</N:3> = New IntPtr(1)
```
Comment (tmat) [inline src/Compilers/VisualBasic/Test/Emit/Emit/EditAndContinue/EditAndContinueTests.vb]: >Dim <N:3>d</N:3> = New IntPtr(1) [](start = 8, length = 32)

VB doesn't have pointers. We can remove this case.

Comment (ivanbasov) [inline src/Compilers/VisualBasic/Test/Emit/Emit/EditAndContinue/EditAndContinueTests.vb]: OK. Thanks!

---
In reply to: [116093741](https://github.com/dotnet/roslyn/pull/19283#discussion_r116093741) [](ancestors = 116093741)

File: src/Compilers/CSharp/Test/Emit/Emit/EditAndContinue/EditAndContinueTests.cs (hunk 2 above)
Comment (ivanbasov) [inline src/Compilers/CSharp/Test/Emit/Emit/EditAndContinue/EditAndContinueTests.cs]: Thank you! This is done

---
In reply to: [116093049](https://github.com/dotnet/roslyn/pull/19283#discussion_r116093049) [](ancestors = 116093049)

File: src/Compilers/VisualBasic/Test/Emit/Emit/EditAndContinue/EditAndContinueTests.vb (hunk 3 above)
Comment (ivanbasov) [inline src/Compilers/VisualBasic/Test/Emit/Emit/EditAndContinue/EditAndContinueTests.vb]: Thank you! Added a test and made similar changes to VB. An existing VB test fails.

---
In reply to: [116093622](https://github.com/dotnet/roslyn/pull/19283#discussion_r116093622) [](ancestors = 116093622)

**Comment by ivanbasov** [comment]
Thank you for the feedback, @tmat ! I made corresponding changes. As I said, an existing VB test failed after changes in VisualBasicSymbolMatcher.vb

**Comment by ivanbasov** [comment]
E2E tests in VB with tuples, e.g. reordering foreach over tuples work fine

File: src/Compilers/VisualBasic/Portable/Emit/EditAndContinue/VisualBasicSymbolMatcher.vb [hunk 5]
```diff
@@ -692,6 +692,11 @@ Namespace Microsoft.CodeAnalysis.VisualBasic.Emit
             End Function
 
             Public Overrides Function VisitNamedType(type As NamedTypeSymbol) As Symbol
+                If (type.IsTupleType) Then
```
Comment (tmat) [inline src/Compilers/VisualBasic/Portable/Emit/EditAndContinue/VisualBasicSymbolMatcher.vb]: >( [](start = 19, length = 1)

nit: No parenthesis needed in VB

File: src/Compilers/CSharp/Portable/Emitter/EditAndContinue/CSharpSymbolMatcher.cs [hunk 6]
```diff
@@ -498,7 +498,7 @@ public override Symbol VisitNamedType(NamedTypeSymbol sourceType)
                         return null;
                     }
 
-                    return TupleTypeSymbol.Create(otherDef, sourceType.TupleElementNames);
+                    return otherDef;
```
Comment (jcouv) [inline src/Compilers/CSharp/Portable/Emitter/EditAndContinue/CSharpSymbolMatcher.cs]: Consider adding a comment.

File: src/Compilers/CSharp/Test/Emit/Emit/EditAndContinue/LocalSlotMappingTests.cs [hunk 7]
```diff
@@ -3521,5 +3521,204 @@ .maxstack  2
...
+class C
+{
+    public unsafe static void G()
+    {        
+        var <N:0>a</N:0> = new { key = ""a"", value = new List<(int, int)>()};
+        var <N:1>b</N:1> = (number: 5, value: a);
+        var <N:2>c</N:2> = new[] { b };
+        int[] <N:3>array</N:3> = { 1, 2, 3 };
+        ref int <N:4>d</N:4> = ref array[0];
+        C1<(int, dynamic)>.E***[,,] <N:5>x</N:5> = null;
```
Comment (jcouv) [inline src/Compilers/CSharp/Test/Emit/Emit/EditAndContinue/LocalSlotMappingTests.cs]: Whoa

**Comment by jcouv** [comment]
The PR title mentions tuples in foreach, but I see no such test in the diff. I'm confused.

File: src/Compilers/CSharp/Test/Emit/Emit/EditAndContinue/LocalSlotMappingTests.cs [hunk 8]
```diff
@@ -3521,5 +3521,204 @@ .maxstack  2
...
+}
+");
+
+            var md0 = ModuleMetadata.CreateFromImage(v0.EmittedAssemblyData);
+
+            var generation0 = EmitBaseline.CreateInitialBaseline(md0, v0.CreateSymReader().GetEncMethodDebugInfo);
+            var diff1 = compilation1.EmitDifference(
+                generation0,
+                ImmutableArray.Create(
+                    new SemanticEdit(SemanticEditKind.Update, f0, f1, GetSyntaxMapFromMarkers(source0, source1), preserveLocalVariables: true)));
```
Comment (jcouv) [inline src/Compilers/CSharp/Test/Emit/Emit/EditAndContinue/LocalSlotMappingTests.cs]: I don't understand what is the edit. Aren't `f0` and `f1` the same (both use the same source)?

Comment (tmat) [inline src/Compilers/CSharp/Test/Emit/Emit/EditAndContinue/LocalSlotMappingTests.cs]: There is no edit. The compiler doesn't care about edits, only how nodes that declare variables map to previous syntax. We just want to verify that the local slots are assigned correctly.

File: src/Compilers/VisualBasic/Portable/Emit/EditAndContinue/VisualBasicSymbolMatcher.vb (hunk 5 above)
Comment (jcouv) [inline src/Compilers/VisualBasic/Portable/Emit/EditAndContinue/VisualBasicSymbolMatcher.vb]: From our discussion, let's figure out why this didn't result in the extra locals (from previous PR and tests) to be optimized away.

Comment (ivanbasov) [inline src/Compilers/VisualBasic/Portable/Emit/EditAndContinue/VisualBasicSymbolMatcher.vb]: thanks!

---
In reply to: [116131000](https://github.com/dotnet/roslyn/pull/19283#discussion_r116131000) [](ancestors = 116131000)

**Comment by ivanbasov** [comment]
Thank you! Fixed

---
In reply to: [301147392](https://github.com/dotnet/roslyn/pull/19283#issuecomment-301147392) [](ancestors = 301147392)

File: src/Compilers/VisualBasic/Portable/Emit/EditAndContinue/VisualBasicSymbolMatcher.vb (hunk 5 above)
Comment (ivanbasov) [inline src/Compilers/VisualBasic/Portable/Emit/EditAndContinue/VisualBasicSymbolMatcher.vb]: the original question was in [https://github.com/dotnet/roslyn/pull/19114](https://github.com/dotnet/roslyn/pull/19114)

re-reviewing it now

---
In reply to: [116339869](https://github.com/dotnet/roslyn/pull/19283#discussion_r116339869) [](ancestors = 116339869)

Comment (ivanbasov) [inline src/Compilers/VisualBasic/Portable/Emit/EditAndContinue/VisualBasicSymbolMatcher.vb]: I have reviewed scenarios mentioned in the discussion from the past code review [https://github.com/dotnet/roslyn/pull/19114](https://github.com/dotnet/roslyn/pull/19114)
There are EditAndContinueTests.cs with tuples in the foreach statement. We attempted to change their names or types.

1. We still do not have a match for them, i.e. extra locals appear.
2. We still have a rude edit warning for this situation. This prevents us from run-time errors.
3. We already covered the situation with [https://github.com/dotnet/roslyn/issues/19154](https://github.com/dotnet/roslyn/issues/19154)

---
In reply to: [116639124](https://github.com/dotnet/roslyn/pull/19283#discussion_r116639124) [](ancestors = 116639124,116339869)

File: src/Compilers/CSharp/Test/Emit/Emit/EditAndContinue/SymbolMatcherTests.cs [hunk 9]
```diff
@@ -769,7 +769,7 @@ public class C
     public delegate (int, int) F();
 }";
             var source1 = @"
-public struct C
```
Comment (tmat) [inline src/Compilers/CSharp/Test/Emit/Emit/EditAndContinue/SymbolMatcherTests.cs]: >struct [](start = 7, length = 6)

Why this change?

Comment (ivanbasov) [inline src/Compilers/CSharp/Test/Emit/Emit/EditAndContinue/SymbolMatcherTests.cs]: We should compare class vs. class here not class vs. struct. The change happens in the delegate not in the class/struct.

---
In reply to: [116844195](https://github.com/dotnet/roslyn/pull/19283#discussion_r116844195) [](ancestors = 116844195)

Comment (tmat) [inline src/Compilers/CSharp/Test/Emit/Emit/EditAndContinue/SymbolMatcherTests.cs]: Oh, yes. Sorry I missed that.

---
In reply to: [116845747](https://github.com/dotnet/roslyn/pull/19283#discussion_r116845747) [](ancestors = 116845747,116844195)

File: src/Compilers/CSharp/Portable/Emitter/EditAndContinue/CSharpSymbolMatcher.cs (hunk 6 above)
Comment (ivanbasov) [inline src/Compilers/CSharp/Portable/Emitter/EditAndContinue/CSharpSymbolMatcher.cs]: I could not find an idea for the comment. It seems like a simplification change. Do you have ideas for the comment?

---
In reply to: [116290923](https://github.com/dotnet/roslyn/pull/19283#discussion_r116290923) [](ancestors = 116290923)

**Comment by ivanbasov** [comment]
need to resolve merge conflict

**Comment by tmat** [comment]
@dotnet/roslyn-compiler Need 2 reviews please.

File: src/Compilers/VisualBasic/Portable/Symbols/ArrayTypeSymbol.vb [hunk 10]
```diff
@@ -317,6 +317,41 @@ Namespace Microsoft.CodeAnalysis.VisualBasic.Symbols
             Return IsSameType(obj, TypeCompareKind.ConsiderEverything)
         End Function
 
+        Friend Overrides Function Equals(t2 As TypeSymbol, comparison As TypeCompareKind) As Boolean
```
Comment (tmat) [inline src/Compilers/VisualBasic/Portable/Symbols/ArrayTypeSymbol.vb]: >Equals [](start = 34, length = 6)

Each of these implementations could use some unit tests. Presumably there are some C# tests that can be ported.

Comment (tmat) [inline src/Compilers/VisualBasic/Portable/Symbols/ArrayTypeSymbol.vb]: Specifically, types with custom modifiers. 

---
In reply to: [117133141](https://github.com/dotnet/roslyn/pull/19283#discussion_r117133141) [](ancestors = 117133141)

File: src/Compilers/VisualBasic/Portable/Symbols/NamedTypeSymbol.vb [hunk 11]
```diff
@@ -448,6 +449,108 @@ Namespace Microsoft.CodeAnalysis.VisualBasic.Symbols
...
+                Return False
+            End If
+
+            If (comparison And TypeCompareKind.IgnoreDynamic) <> 0 Then
+                If t2.TypeKind = TypeKind.Dynamic Then
+                    If Me.SpecialType = SpecialType.System_Object Then
+                        Return True
+                    End If
+                End If
+            End If
```
Comment (tmat) [inline src/Compilers/VisualBasic/Portable/Symbols/NamedTypeSymbol.vb]: No applicable in VB. VB doesn't have Dynamic type.

Comment (ivanbasov) [inline src/Compilers/VisualBasic/Portable/Symbols/NamedTypeSymbol.vb]: Thank you! Fixed.

---
In reply to: [117133293](https://github.com/dotnet/roslyn/pull/19283#discussion_r117133293) [](ancestors = 117133293)

**Comment by AlekseyTs** [comment]
As we discussed offline, we should switch to using IsSameType extension method, rather than adding new Equals overload.

**Comment by ivanbasov** [comment]
Thank you! Fixed

---
In reply to: [302563607](https://github.com/dotnet/roslyn/pull/19283#issuecomment-302563607) [](ancestors = 302563607)

File: src/Compilers/VisualBasic/Portable/Symbols/ArrayTypeSymbol.vb (hunk 10 above)
Comment (ivanbasov) [inline src/Compilers/VisualBasic/Portable/Symbols/ArrayTypeSymbol.vb]: Removed Symbol changes. Current test coverage should be enough. Thank you!

---
In reply to: [117133247](https://github.com/dotnet/roslyn/pull/19283#discussion_r117133247) [](ancestors = 117133247,117133141)

File: src/Compilers/VisualBasic/Portable/Emit/EditAndContinue/VisualBasicSymbolMatcher.vb [hunk 12]
```diff
@@ -652,10 +652,10 @@ Namespace Microsoft.CodeAnalysis.VisualBasic.Emit
                 End Sub
 
                 Public Overloads Function Equals(source As TypeSymbol, other As TypeSymbol) As Boolean
-                    Dim visitedSource = _matcher.Visit(source)
-                    Dim visitedOther = If(_deepTranslatorOpt IsNot Nothing, _deepTranslatorOpt.Visit(other), other)
+                    Dim visitedSource = DirectCast(_matcher.Visit(source), TypeSymbol)
+                    Dim visitedOther = If(_deepTranslatorOpt IsNot Nothing, DirectCast(_deepTranslatorOpt.Visit(other), TypeSymbol), other)
 
-                    Return visitedSource = visitedOther
+                    Return visitedSource IsNot Nothing AndAlso visitedSource.IsSameType(visitedOther, TypeCompareKind.IgnoreTupleNames)
```
Comment (AlekseyTs) [inline src/Compilers/VisualBasic/Portable/Emit/EditAndContinue/VisualBasicSymbolMatcher.vb]: >visitedSource.IsSameType(visitedOther, TypeCompareKind.IgnoreTupleNames) [](start = 63, length = 72)

We are checking if visitedSource is Nothing, it looks like we should do the same check for visitedOther.

**Comment by AlekseyTs** [comment]
Done with review pass.

File: src/Compilers/VisualBasic/Portable/Emit/EditAndContinue/VisualBasicSymbolMatcher.vb (hunk 12 above)
Comment (ivanbasov) [inline src/Compilers/VisualBasic/Portable/Emit/EditAndContinue/VisualBasicSymbolMatcher.vb]: Thank you! Fixed


---
In reply to: [117395320](https://github.com/dotnet/roslyn/pull/19283#discussion_r117395320) [](ancestors = 117395320)

File: src/Compilers/VisualBasic/Portable/Emit/EditAndContinue/VisualBasicSymbolMatcher.vb [hunk 13]
```diff
@@ -652,10 +652,11 @@ Namespace Microsoft.CodeAnalysis.VisualBasic.Emit
                 End Sub
 
                 Public Overloads Function Equals(source As TypeSymbol, other As TypeSymbol) As Boolean
-                    Dim visitedSource = _matcher.Visit(source)
-                    Dim visitedOther = If(_deepTranslatorOpt IsNot Nothing, _deepTranslatorOpt.Visit(other), other)
+                    Dim visitedSource = DirectCast(_matcher.Visit(source), TypeSymbol)
+                    Dim visitedOther = If(_deepTranslatorOpt IsNot Nothing, DirectCast(_deepTranslatorOpt.Visit(other), TypeSymbol), other)
 
-                    Return visitedSource = visitedOther
+                    ' If both visitedSource And visitedOther are Nothing, return false meaning that the method was not able to verify the equality.
```
Comment (AlekseyTs) [inline src/Compilers/VisualBasic/Portable/Emit/EditAndContinue/VisualBasicSymbolMatcher.vb]: >And [](start = 44, length = 3)

"And" vs "and".

File: src/Compilers/VisualBasic/Portable/Emit/EditAndContinue/VisualBasicSymbolMatcher.vb (hunk 12 above)
Comment (ivanbasov) [inline src/Compilers/VisualBasic/Portable/Emit/EditAndContinue/VisualBasicSymbolMatcher.vb]: Thank you!

---
In reply to: [117552926](https://github.com/dotnet/roslyn/pull/19283#discussion_r117552926) [](ancestors = 117552926,117395320)

File: src/Compilers/VisualBasic/Test/Emit/CodeGen/CodeGenTuples.vb [hunk 14]
```diff
@@ -14803,7 +14803,8 @@ options:=TestOptions.DebugExe, additionalRefs:=s_valueTupleRefs)
         End Sub
 
         Private Shared Sub AssertTestDisplayString(symbols As ImmutableArray(Of Symbol), ParamArray baseLine As String())
-            AssertEx.Equal(symbols.Select(Function(s) s.ToTestDisplayString()), baseLine)
+            ' Re-ordering arguments because expected Is usually goes first.
+            AssertEx.Equal(baseLine, symbols.Select(Function(s) s.ToTestDisplayString()))
```
Comment (cston) [inline src/Compilers/VisualBasic/Test/Emit/CodeGen/CodeGenTuples.vb]: `is` rather than `Is`, and perhaps just `expected is first`.

Comment (ivanbasov) [inline src/Compilers/VisualBasic/Test/Emit/CodeGen/CodeGenTuples.vb]: Thank you! Fixed

---
In reply to: [117559468](https://github.com/dotnet/roslyn/pull/19283#discussion_r117559468) [](ancestors = 117559468)

File: src/Compilers/VisualBasic/Portable/Emit/EditAndContinue/VisualBasicSymbolMatcher.vb (hunk 13 above)
Comment (ivanbasov) [inline src/Compilers/VisualBasic/Portable/Emit/EditAndContinue/VisualBasicSymbolMatcher.vb]: Thank you! Fixed

---
In reply to: [117555598](https://github.com/dotnet/roslyn/pull/19283#discussion_r117555598) [](ancestors = 117555598)

File: src/Compilers/VisualBasic/Test/Emit/Emit/EditAndContinue/SymbolMatcherTests.vb [hunk 15]
```diff
@@ -512,5 +512,327 @@ End Class
...
+                compilation1.SourceAssembly,
+                New EmitContext(),
+                compilation0.SourceAssembly,
+                New EmitContext(),
+                Nothing)
+
+            Dim member = compilation1.GetMember(Of FieldSymbol)("C.x")
+            Dim other = matcher.MapDefinition(member)
+            ' Types must match because just an element name was changed.
+            Assert.NotNull(other)
```
Comment (cston) [inline src/Compilers/VisualBasic/Test/Emit/Emit/EditAndContinue/SymbolMatcherTests.vb]: Perhaps `Assert.Equal("... signature ...", other.ToTestDisplayString())`

File: src/Compilers/VisualBasic/Test/Emit/Emit/EditAndContinue/SymbolMatcherTests.vb [hunk 16]
```diff
@@ -512,5 +512,327 @@ End Class
...
+                New EmitContext(),
+                compilation0.SourceAssembly,
+                New EmitContext(),
+                Nothing)
+
+            Dim member = compilation1.GetMember(Of SourceNamedTypeSymbol)("C.F")
+            Dim other = matcher.MapDefinition(member)
+            ' Tuple delegate defines a type. We should be able to match old and new types by name.
+            Assert.NotNull(other)
+        End Sub
```
Comment (cston) [inline src/Compilers/VisualBasic/Test/Emit/Emit/EditAndContinue/SymbolMatcherTests.vb]: Consider testing type changes from non-tuple to tuple and tuple to non-tuple.

File: src/Compilers/VisualBasic/Test/Emit/Emit/EditAndContinue/SymbolMatcherTests.vb (hunk 15 above)
Comment (ivanbasov) [inline src/Compilers/VisualBasic/Test/Emit/Emit/EditAndContinue/SymbolMatcherTests.vb]: Good idea! Thank you! Done

---
In reply to: [117564324](https://github.com/dotnet/roslyn/pull/19283#discussion_r117564324) [](ancestors = 117564324)

File: src/Compilers/VisualBasic/Test/Emit/Emit/EditAndContinue/SymbolMatcherTests.vb (hunk 16 above)
Comment (ivanbasov) [inline src/Compilers/VisualBasic/Test/Emit/Emit/EditAndContinue/SymbolMatcherTests.vb]: Thank you! Added

---
In reply to: [117565093](https://github.com/dotnet/roslyn/pull/19283#discussion_r117565093) [](ancestors = 117565093)

**Comment by cston** [comment]
LGTM

**Comment by ivanbasov** [comment]
Tagging @MattGertz or @MeiChin-Tsai for ask mode or escrow approval

**Comment by MattGertz** [comment]
OK by me.  Once Mei-Chin approves, please go ahead & set the label.

**Comment by MeiChin-Tsai** [comment]
approved.
